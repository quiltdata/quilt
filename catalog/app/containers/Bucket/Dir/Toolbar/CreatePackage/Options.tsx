import * as React from 'react'
import * as M from '@material-ui/core'
import * as Lab from '@material-ui/lab'

import { EmptySlot, ErrorSlot } from 'containers/Bucket/Successors'
import * as Request from 'utils/useRequest'
import * as workflows from 'utils/workflows'

const LIST_ITEM_TYPOGRAPHY_PROPS = { noWrap: true }

const useStyles = M.makeStyles((t) => ({
  root: {
    minWidth: t.spacing(40),
  },
}))

interface CreatePackageOptionsProps {
  onChange: (x: workflows.Successor) => void
  successors: Request.Result<workflows.Successor[]>
}

export default function CreatePackageOptions({
  onChange,
  successors,
}: CreatePackageOptionsProps) {
  const classes = useStyles()
  if (successors === Request.Idle) return null

  if (successors instanceof Error) {
    return (
      <div className={classes.root}>
        <ErrorSlot error={successors} />
      </div>
    )
  }

  if (successors === Request.Loading) {
    return (
      <M.List dense className={classes.root}>
        <M.ListItem>
          <M.ListItemText primary={<Lab.Skeleton />} secondary={<Lab.Skeleton />} />
        </M.ListItem>
        <M.ListItem>
          <M.ListItemText primary={<Lab.Skeleton />} secondary={<Lab.Skeleton />} />
        </M.ListItem>
        <M.ListItem>
          <M.ListItemText primary={<Lab.Skeleton />} secondary={<Lab.Skeleton />} />
        </M.ListItem>
      </M.List>
    )
  }

  if (!successors.length) {
    return (
      <div className={classes.root}>
        <EmptySlot />
      </div>
    )
  }

  return (
    <M.List dense className={classes.root}>
      {successors.map((successor) => (
        <M.ListItem
          key={successor.slug}
          onClick={(event) => {
            event.stopPropagation()
            onChange(successor)
          }}
          button
        >
          <M.ListItemText
            primary={successor.name}
            primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
            secondary={successor.url}
            secondaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
          />
        </M.ListItem>
      ))}
    </M.List>
  )
}
