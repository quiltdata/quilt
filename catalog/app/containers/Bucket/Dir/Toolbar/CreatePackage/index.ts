export { default as Options } from './Options'
export { default as useSuccessors } from './useSuccessors'
