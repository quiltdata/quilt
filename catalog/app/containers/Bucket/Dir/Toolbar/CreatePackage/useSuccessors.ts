import * as React from 'react'

import { workflowsConfig } from 'containers/Bucket/requests'
import * as AWS from 'utils/AWS'
import * as Request from 'utils/useRequest'
import { bucketToSuccessor } from 'utils/workflows'
import type { Successor } from 'utils/workflows'

export default function useSuccessors(bucket: string): Request.Result<Successor[]> {
  const s3 = AWS.S3.use()
  const req = React.useCallback(() => workflowsConfig({ s3, bucket }), [bucket, s3])
  const data = Request.use(req)

  if (data instanceof Error || data === Request.Idle || data === Request.Loading) {
    return data
  }

  return data.successors.find(({ slug }) => slug === bucket)
    ? data.successors
    : [bucketToSuccessor(bucket), ...data.successors]
}
