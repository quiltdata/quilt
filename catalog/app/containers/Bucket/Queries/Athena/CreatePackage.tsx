import * as React from 'react'
import * as M from '@material-ui/core'

import * as AddToPackage from 'containers/AddToPackage'
import { usePackageCreationDialog } from 'containers/Bucket/PackageDialog/PackageCreationForm'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

import * as requests from '../requests'

type ManifestKey = 'hash' | 'logical_key' | 'meta' | 'physical_keys' | 'size'
type ManifestEntryStringified = Record<ManifestKey, string>

function SeeDocsForCreatingPackage() {
  return (
    <M.Tooltip title="You can create packages from the query results. Click to see the docs.">
      <a href="https://docs.quiltdata.com/advanced/athena" target="_blank">
        <M.IconButton>
          <M.Icon>help_outline</M.Icon>
        </M.IconButton>
      </a>
    </M.Tooltip>
  )
}

function doQueryResultsContainManifestEntries(
  rows: string[][],
): rows is [ManifestKey[], ...string[][]] {
  const [head] = rows
  return (
    head.includes('size') &&
    head.includes('physical_keys') &&
    head.includes('logical_key')
  )
}

function rowToManifestEntryStringified(
  row: string[],
  head: ManifestKey[],
): ManifestEntryStringified {
  return row.reduce((acc, value, index) => {
    if (!head[index]) return acc
    return {
      ...acc,
      [head[index]]: value,
    }
  }, {} as ManifestEntryStringified)
}

function parseManifestEntryStringified(entry: ManifestEntryStringified): {
  [key: string]: Model.S3File
} | null {
  if (!entry.logical_key) return null
  if (!entry.physical_keys) return null
  try {
    const handle = s3paths.parseS3Url(
      entry.physical_keys.replace(/^\[/, '').replace(/\]$/, ''),
    )
    const sizeParsed = Number(entry.size)
    const size = Number.isNaN(sizeParsed) ? 0 : sizeParsed
    return {
      [entry.logical_key]: {
        ...handle,
        size,
      },
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e)
    return null
  }
}

function parseQueryResults(
  rows: [ManifestKey[], ...string[][]],
): Record<string, Model.S3File> {
  const [head, ...tail] = rows
  const manifestEntries: ManifestEntryStringified[] = tail.reduce(
    (memo, row) => memo.concat(rowToManifestEntryStringified(row, head)),
    [] as ManifestEntryStringified[],
  )
  return manifestEntries.reduce(
    (memo, entry) => ({
      ...memo,
      ...parseManifestEntryStringified(entry),
    }),
    {},
  )
}

interface CreatePackageProps {
  bucket: string
  rows: requests.athena.QueryResultsRows
}

export default function CreatePackage({ bucket, rows }: CreatePackageProps) {
  const addToPackage = AddToPackage.use()
  const createDialog = usePackageCreationDialog({
    bucket,
    delayHashing: true,
    disableStateDisplay: true,
  })
  const onPackage = React.useCallback(() => {
    if (!doQueryResultsContainManifestEntries(rows)) return

    // TODO: make it lazy, and disable button
    const entries = parseQueryResults(rows)
    addToPackage?.merge(entries)
    createDialog.open()
  }, [addToPackage, createDialog, rows])

  if (!doQueryResultsContainManifestEntries(rows)) return <SeeDocsForCreatingPackage />

  return (
    <>
      {createDialog.render({
        successTitle: 'Package created',
        successRenderMessage: ({ packageLink }) => (
          <>Package {packageLink} successfully created</>
        ),
        title: 'Create package',
      })}
      <M.Button color="primary" onClick={onPackage} size="small" variant="outlined">
        Create package
      </M.Button>
    </>
  )
}
