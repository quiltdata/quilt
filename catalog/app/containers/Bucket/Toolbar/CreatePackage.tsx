import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import { useSelection } from 'containers/Bucket/Selection/Provider'

const useStyles = M.makeStyles({
  root: {
    transitionProperty: 'background-color, box-shadow', // Disable border-width transition
  },
})

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  label?: string
}

export default function Button({ className, ...props }: ButtonProps) {
  const classes = useStyles()
  const slt = useSelection()
  return (
    <Buttons.WithPopover
      className={cx(classes.root, className)}
      label="Create package"
      variant={slt.isEmpty ? 'outlined' : 'contained'}
      color={slt.isEmpty ? 'default' : 'primary'}
      {...props}
    />
  )
}
