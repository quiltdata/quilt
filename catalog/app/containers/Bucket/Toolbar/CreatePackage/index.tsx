import * as React from 'react'
import * as Buttons from 'components/Buttons'

import { useSelection } from '../../Selection/Provider'

export { default as BucketDirOptions } from './BucketDirOptions'
export { default as useSuccessors } from './useSuccessors'

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  label?: string
}

export function Button(props: ButtonProps) {
  const slt = useSelection()
  // FIXME: if slt == null => File button
  // FIXME: make WithPopover#icon optional
  return (
    <Buttons.WithPopover
      label="Create package"
      variant={slt.isEmpty ? 'outlined' : 'contained'}
      color={slt.isEmpty ? 'default' : 'primary'}
      {...props}
    />
  )
}
