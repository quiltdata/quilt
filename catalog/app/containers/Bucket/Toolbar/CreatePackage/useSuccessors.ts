import * as React from 'react'

import * as AWS from 'utils/AWS'
import * as Request from 'utils/useRequest'
import type { Successor } from 'utils/workflows'

import { workflowsConfig } from '../../requests'

export default function useSuccessors(bucket: string): Request.Result<Successor[]> {
  const s3 = AWS.S3.use()
  const req = React.useCallback(() => workflowsConfig({ s3, bucket }), [bucket, s3])
  const data = Request.use(req)

  if (data instanceof Error || data === Request.Idle || data === Request.Loading) {
    return data
  }

  return data.successors
}
