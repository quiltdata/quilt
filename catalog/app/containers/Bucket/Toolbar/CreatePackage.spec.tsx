import * as React from 'react'
import renderer from 'react-test-renderer'

import CreatePackage from './CreatePackage'

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  IconButton: ({ className, children }: any) => <button {...{ className, children }} />,
  Button: ({ className, children }: any) => <button {...{ className, children }} />,
}))

jest.mock('containers/Bucket/Selection/Provider', () => ({
  useSelection: jest.fn(() => ({
    isEmpty: true,
    totalCount: 0,
    inited: true,
  })),
}))

describe('containers/Bucket/Toolbar/CreatePackage', () => {
  it('should render with empty selection', () => {
    const tree = renderer
      .create(<CreatePackage className="custom-class">Hello, Popover!</CreatePackage>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
