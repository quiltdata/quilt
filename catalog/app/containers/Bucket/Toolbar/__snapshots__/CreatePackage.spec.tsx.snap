// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Toolbar/CreatePackage should render with empty selection 1`] = `
<div
  className="makeStyles-root-2"
>
  <button
    className="makeStyles-root-1 custom-class"
  >
    Create package
  </button>
  <div
    aria-hidden={true}
    className="MuiBackdrop-root makeStyles-backdrop-4"
    onClick={[Function]}
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  />
</div>
`;
